### DISCUSSION TITLE
Is there any possibility to have graph-support as an embedded mini-graphviz?

### DISCUSSION THREAD

**Discussion by jamisonjiang** [plantuml/plantuml#1383] [Ideas]
I learned that PlantUML has tried various ways to embed graphviz (like Smetana, Jdot), and PlantUML may only rely on a small part of the dot layout-related algorithms. 
[graph-support](https://github.com/jamisonjiang/graph-support)  is a tiny embedded graphviz developed. I am more interested in whether its functions are sufficient to support the needs of PlantUML.

**Comment by arnaudroques** [comment]
First, I'd like to say that we are very impressed by your work.
Re-implementing graphviz in Java is definitively a hard task and it looks like you did a much more greater code that ours!

So yes, I think that [graph-support](https://github.com/jamisonjiang/graph-support) _could_ be used with PlantUML.

However, I think it would be much more effective to define a new API that would allow to:

- separate the layout part from the drawing part (mainly because the drawing part is done by PlantUML)
- add new features that are not supported because of limitation of the `dot` language.

[graph-support](https://github.com/jamisonjiang/graph-support) could support both the actual API like dot-script and a new one.

Just to give you an idea of what _could_ be that new API:

```
GraphSupport graph = GraphSupport.digraph();

// Add a new 100 x 200 node
Node nd_1 = graph.nodeBuilder().rectangle(100, 200).build();

// Add a new 200 x 300 node
Node nd_2 = graph.nodeBuilder().rectangle(100, 300).build();

// Build a line with minlen = 2
Line line1_2 = graph.addLine(nd_1, nd_2, graph.lineProperties().minlen(2));

graph.doLayout();

// Get node1 position, not sure about the result
Object position1 = nd_1.getPosition();

// Get node1 position, not sure about the result
Object position2 = nd_2.getPosition();

// Get line geometry, not sure about the result
Object line1 = line1_2.getCurves();
```

Of course, this is really a simple and wild draft.

If you are interested, we could work together: 
- we define together the API in a way that is both usable for PlantUML and implementable for [graph-support](https://github.com/jamisonjiang/graph-support)
- we could write some unitary tests to help you to code your library
- on our side, we add this new layout engine to PlantUML

This way, I think we could get a generic layout library which could be used by many other Java projects.

Any though?

**Comment by jamisonjiang** [comment]
Sure,I am interested in combining related work. If it can be combined with **PlantUML**, it proves that this library is stable and efficient. I reply to your content one by one:
### Separate the layout
Now rendering and layout are indeed separated, you can skip rendering and only perform layout, and then get the node coordinates and each point of the edge.(There are related examples at the end of the README)
### Support the `dot` script
I focused on the development of graphviz-related features, and did not spend energy developing a dot parser. But if there is a dot parser, combined with graph-support is very fast. 
It took me 20 minutes to integrate [JPGD](http://www.alexander-merz.com/graphviz/) in my local test, so as to test graph-support with dot script. Unfortunately, the structure of [JPGD](http://www.alexander-merz.com/graphviz/) parsing is relatively limited, and some commonly used dot writing methods may fail to parse, so I don't plan to use it as the default parser of graph-support, and it can only be used for testing at present. 
At present, the Java version does not have a better choice than [JPGD](http://www.alexander-merz.com/graphviz/), so if I want to support this, I need to redevelop it from scratch. There hava a lot of work. I am not sure whether this will be supported in the future. It depends on whether it is A common need.
### New api proposal
The suggestion of the new api is good. A similar effect should be achieved through simple encapsulation. Maybe we can cooperate to encapsulate what kind of api can quickly integrate two applications.Then for the current API, an example to achieve a similar effect is as follows:
```java
    // Add a new 100 x 200 node
    Node nd_1 = Node.builder()
        .shape(NodeShapeEnum.RECT)
        .height((double) 100 / Graphviz.PIXEL)
        .width((double) 200 / Graphviz.PIXEL)
        .build();

    // Add a new 200 x 300 node
    Node nd_2 = Node.builder()
        .shape(NodeShapeEnum.RECT)
        .height((double) 200 / Graphviz.PIXEL)
        .width((double) 300 / Graphviz.PIXEL)
        .build();

    // Build a line with minlen = 2
    Line line1_2 = Line.builder(nd_1, nd_2).minlen(2).build();

    // Layout, Might be a bit more complicated, some more code than your example
    LayoutEngine layoutEngine = Layout.DOT.getLayoutEngine();
    // This step does not trigger rendering, only layout
    DrawGraph graph = layoutEngine.layout(Graphviz.digraph().addLine(line1_2).build());

    // Get node1 position
    NodeDrawProp position1  = graph.getNodeDrawProp(nd_1);

    // Get node2 position
    NodeDrawProp position2  = graph.getNodeDrawProp(nd_1);

    // Get line geometry
    LineDrawProp line1 = graph.getLineDrawProp(line1_2);
```
This looks very similar to your example, without much more code.


Finally, I'm not sure how **PlantUML** does the **Graphviz** integration work, and if there's anything in the design of **graph-support** that indicates integration is indeed not possible (similar to how PlantUML must parse dot scripts), I'll try to see how to fix it..

**Reply by evantill** [reply]
Impressive work infact. Congratulations ! 

Just a question on classes in `org.apache_gs` package : Is this a customized version or just a verbatim copy of the apache common project ?

**Reply by jamisonjiang** [reply]
> Impressive work infact. Congratulations !
> 
> Just a question on classes in `org.apache_gs` package : Is this a customized version or just a verbatim copy of the apache common project ?

Copied a small part from apache common for escaping html tags. The renaming is to avoid the package name affecting the complete apache common referenced by the relying party.

**Reply by evantill** [reply]
I have security concerns about it: escaping is a _sensitive topic_ (in my humble opinion).

The simplest solution would be to declare dependencies.

A middle ground solution could be to use the [Maven Shade Plugin](https://maven.apache.org/plugins/maven-shade-plugin/)
> See this [stackoverflow answer](https://stackoverflow.com/a/19386712) for a good example on how to use it.

**Reply by jamisonjiang** [reply]
> I have security concerns about it: escaping is a _sensitive topic_ (in my humble opinion).
> 
> The simplest solution would be to declare dependencies.
> 
> A middle ground solution could be to use the [Maven Shade Plugin](https://maven.apache.org/plugins/maven-shade-plugin/)
> 
> > See this [stackoverflow answer](https://stackoverflow.com/a/19386712) for a good example on how to use it.

Shade is a good choice, but my goal is to remove all dependencies that can be removed (except the log framework), and even remove the dependencies related to java.awt to ensure that the project can run in the Android environment. This is how I started this project The goal I set for myself is also my technical "aesthetic" (lightweight).
Apache Common is already very light, but unfortunately I still feel that it is too heavy. In fact, I only use a few classes, but I have to introduce a large dependency. Considering that the protocol of the apache license is compatible with my current practice, I chose to copy the corresponding class files. If there are security and legal issues, I will choose to re-implement it myself. In fact, this part of the work is not difficult, but I was a little lazy at the time.

**Reply by evantill** [reply]
no legal issue IMHO

**Comment by arnaudroques** [comment]
> This looks very similar to your example, without much more code.

Yes, that's fine

> Finally, I'm not sure how PlantUML does the Graphviz integration work, and if there's anything in the design of graph-support that indicates integration is indeed not possible (similar to how PlantUML must parse dot scripts), I'll try to see how to fix it..

Well, the integration of Graphviz is really a nasty hack, something we don't want to reproduce.
Anyway, I can tell you how it works:

Give [this very simple example](http://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhEIImkLWWjJYrIgEPIKCXBp4q5KeafHQd5nLorN5nW6LrTEmMtp9nK1Ik5dCIar1n3iB2L5KWrH38_7oPdnCNba9gN0ZGR0000):

```
@startuml
' This directive generates locally the dot file created by PlantUML
!pragma svek_trace on
class User {
  name
  adress
}

User --> File : label1
User::name --> Foo : label2

@enduml
```


In that case, PlantUML generates the following dot file:

```
digraph unix {
nodesep=0.486111;
ranksep=0.833333;
remincross=true;
searchsize=500;
sh0006 [shape=plaintext,label=<<TABLE BGCOLOR="#000006" BORDER="0" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0"><TR><TD  FIXEDSIZE="TRUE" WIDTH="64.0" HEIGHT="36.0"></TD></TR><TR><TD  FIXEDSIZE="TRUE" WIDTH="64.0" HEIGHT="16.296875" PORT="pb068931cc450442b63f5b3d276ea4297"></TD></TR><TR><TD  FIXEDSIZE="TRUE" WIDTH="64.0" HEIGHT="28.296875"></TD></TR></TABLE>>];

sh0007 [shape=rect,label="",width=0.763889,height=0.666667,color="#000007"];
sh0008 [shape=rect,label="",width=0.805556,height=0.666667,color="#000008"];
sh0006->sh0007[arrowsize=0.5,minlen=1,color="#000009",label=<<TABLE BGCOLOR="#00000A" FIXEDSIZE="TRUE" WIDTH="40" HEIGHT="17"><TR><TD></TD></TR></TABLE>>];
sh0006:pb068931cc450442b63f5b3d276ea4297->sh0008[arrowsize=0.5,minlen=1,color="#00000D",label=<<TABLE BGCOLOR="#00000E" FIXEDSIZE="TRUE" WIDTH="40" HEIGHT="17"><TR><TD></TD></TR></TABLE>>];

}
```

As you see, we use width, height and HTML table to define nodes and labels size.
We also put a specific color for each line, node and label.
Then we launch Graphviz to generate a SVG file.
And we parse this SVG file. Thanks to unique color for each elements, we retrieve geometrical information about each node, label and line.

But, once again, all this is a terrible hack and we certainly don't want to reproduce this :-)


<hr>

So your API is really a good start: we could use it.

However, they are some limitations in Graphviz/Dot.
For example, let's talk about Record/Port.

It would be very nice for us if we could define more precisely some portion/Port inside a node and draw lines between those.

Something like:

```
    // Add a new 100 x 200 node
    Node nd_1 = Node.builder()
        .shape(NodeShapeEnum.RECT)
        .width((double) 100 / Graphviz.PIXEL)
        .height((double) 200 / Graphviz.PIXEL)
        .build();

    // Add a new 200 x 300 node
    Node nd_2 = Node.builder()
        .shape(NodeShapeEnum.RECT)
        .width((double) 200 / Graphviz.PIXEL)
        .height((double) 300 / Graphviz.PIXEL)
        .build();
        
    // Define a "port" 10 x 20 inside nd_1 in the upper left corner
    // The position is relative to the upper left corner
    Port port_1 = nd_1().portBuilder().
        .shape(NodeShapeEnum.RECT)
        .width((double) 10 / Graphviz.PIXEL)
        .height((double) 20 / Graphviz.PIXEL)
        .position(0, 0).
        .build();
    
    // Define a "port" 200 x 10 inside nd_2 at the bottom
    Port port_2 = nd_2().portBuilder().
        .shape(NodeShapeEnum.RECT)
        .width((double) 200 / Graphviz.PIXEL)
        .height((double) 10 / Graphviz.PIXEL)
        .position(0, (double) 290 / Graphviz.PIXEL).
        .build();
   // 290 because it is 300 - 10: so the port is at the bottom of the node.
    
    // Build a line with minlen = 2
    Line line1_2 = Line.builder(port_1, port_2).minlen(2).build();
```

It is just an example to show you what we would like to do.
Does it make sense?
Is it possible to already do that?

**Reply by jamisonjiang** [reply]
Therefore, there are two possible directions to realize this requirement under the current graph-support:
(1) Support dot parsing and also support the complete syntax of table;
pros: PlantUML does not need to be adjusted, and the previous logic can be reused
cons:graph-support needs a lot of work to support these two big features
(2) Use cell expression (similar: "1|{2|3}|4"), then provide some parameters to adjust the size of the port, and provide some simple ways to obtain some geometric properties of the corresponding port.
pros: PlantUML needs to adjust some logic, according to the PlantUML script -> Table syntax needs to be adjusted, and the logic of parsing svg becomes directly calling the api to get the Port result.
cons:graph-support is supported with just a few tweaks, and I can get it done in a day or so.

**Reply by arnaudroques** [reply]
> Therefore, there are two possible directions to realize this requirement under the current graph-support:

Thanks for your answer.
The important point is **under the current graph-support**.

This is only a personal opinion, but to me you should consider drop any link to "dot language".
Don't get me wrong: got/graphviz is a great tool, we use if for more then 10 years.
But it's really old. The concept of "dot language" which was needed at that time can now be replace by a proper API.

In the API, you should not need stuff as "parse cell expression" or "table syntax".

I understand that this is a huge change for the current graph-support and that it would probably need a lot of work on your side.
So it's not a decision for a short terme release, it is really a reflexion on future releases.
But I can tell you as a simple user of layout API, it would be great :-)

**Reply by jamisonjiang** [reply]
Great to hear your thoughts, just this one I wanted to make sure I didn't get it wrong:
> In the API, you should not need stuff as "parse cell expression" or "table syntax".

I'm not sure which of the following you mean:
* Support for these two features is unnecessary?
* These two features do not need to be packaged into Api? Directly need to support the expression of the string structure?

**Reply by arnaudroques** [reply]
> I'm not sure which of the following you mean

Sorry, it's difficult to explain :-)

What I really mean: HTML table support and cell expression ("1|{2|3}|4") **was** just a convenient way to split a node into several "records/ports".
It is good but it is not flexible enough.
Here is a drawing with a single big node in black:
![singlenode](https://user-images.githubusercontent.com/467517/230596834-8acb8d5d-9111-4af4-85b9-db731afc4bd5.png)
I would like to define 7 "records/ports", drawn in red.

In the API, I don't want any `String`, I just would like to define my 7 "records/ports" inside the node this way:

```
    Node nd_1 = Node.builder()
        .shape(NodeShapeEnum.RECT)
        .width(200)
        .height(100)
        .build();

    // createPort(x, y, width, height)
    Port p1 = nd_1.createPort(0, 0, 30, 30)
    Port p2 = nd_1.createPort(15, 30, 40, 300)
    Port p3 = ...
    Port p4 = ...
    Port p5 = ...
    Port p6 = ...
    Port p7 = nd_1.createPort(0, 80, 200, 20)

```

So we don't need any `String` to define the port.
Now I have no idea about internal structures of graph-support, so probably it is much more easy to say than to code.

Do you see what I mean?

**Reply by jamisonjiang** [reply]
Yeah, I see what you mean. Even Html Label has many restrictive rules, and Port cannot be placed at any position of a node. The more basic way is to place Port at any position of a node according to certain rules. Html Label and Cell expressions are just one of the underlying rules "Syntactic sugar". I think this is a really great idea!

One problem that needs to be solved here is that this feature needs to be designed by itself. Because if just implement the features of graphviz, there is only development work. But when a lower-level "rule" is introduced, I need to design another rule so that it can be flexible enough to achieve an effect similar to Html Label. I can try it out and feel free to suggest your ideas too, thanks.

**Reply by arnaudroques** [reply]
> One problem that needs to be solved here is that this feature needs to be designed by itself. 

Yes, exactly!


> But when a lower-level "rule" is introduced, I need to design another rule so that it can be flexible enough to achieve an effect similar to Html Label

Yes, that's fine. I understand that, it definitively make sense.

The fact is that I have several other "crazy" suggestions like this (about labels on edge for example, about cluster margin...)

So, before that you code anything, my initial suggestion is to write to a full documentation about future need into a separate document, so that you can get a full vision of our needs.

We could write together this API specification as some markdown document into [a new folder of graph-support](https://github.com/jamisonjiang/graph-support/api_specification_v2). It will probably take several weeks.
This way, you could get a complete and stable idea about the API and you could start the design afterwards.

Once again, I understand that it means some work on your side, so it's really up to you!
What you think about it?

**Reply by jamisonjiang** [reply]
The original intention of my development of graph-support was that I wanted a lightweight solution using graphviz in Java. I found that I couldn't find related projects, so I did it myself. Therefore, the standard for supporting features in the first version of graph-support is: The features I use most often.

So when the first version is completed, I am not sure which functions to support in the future, because I have lost the standard (personal needs have been met), and the next stage needs to collect other people's needs, and then start next stage work by the priority (because it is impossible to implement the attributes one by one according to the graphviz document).

Now if the integration with PlantUML is the standard for the next stage of development, I will try to implement any functions that help the integration. So if you ask me to say which features I will support in the future, I will say that I don't know what kind of capabilities PlantUML needs. What is currently known is the capability of the Port in this node.

In addition, I reserved some design flexibility when implementing Cell expressions, because I thought in advance whether this design can be reused if Html labels are to be supported in the future. So I have some tentative ideas on how to support the relevant content you described through the existing structure. I have constructed it in my mind and it will not be too complicated. I will post some detailed ideas later.

**Reply by jamisonjiang** [reply]
Can you list in detail which Graphviz features PlantUML uses, and the limitations encountered when using smetana and vizjs. I want to confirm whether graph-support can completely solve these problems, and if so, how much work is needed. Thanks in advance.

**Reply by arnaudroques** [reply]
Sure, we will start a document which lists which Graphviz feature we use. We need several days for this, so please be patient :-)
Don't get me wrong: we will not be able to integrate graph-support on short term because we are already in the middle of too many changes (including our own build).
However, on the long term, this is definitively something (so this integration of graph-support) we have in mind.

**Reply by jamisonjiang** [reply]
> > One problem that needs to be solved here is that this feature needs to be designed by itself.
> 
> Yes, exactly!
> 
> > But when a lower-level "rule" is introduced, I need to design another rule so that it can be flexible enough to achieve an effect similar to Html Label
> 
> Yes, that's fine. I understand that, it definitively make sense.
> 
> The fact is that I have several other "crazy" suggestions like this (about labels on edge for example, about cluster margin...)
> 
> So, before that you code anything, my initial suggestion is to write to a full documentation about future need into a separate document, so that you can get a full vision of our needs.
> 
> We could write together this API specification as some markdown document into [a new folder of graph-support](https://github.com/jamisonjiang/graph-support/api_specification_v2). It will probably take several weeks. This way, you could get a complete and stable idea about the API and you could start the design afterwards.
> 
> Once again, I understand that it means some work on your side, so it's really up to you! What you think about it?

Please see this document https://github.com/jamisonjiang/graph-support/wiki/Assemble-document

**Reply by arnaudroques** [reply]
Sounds like a good start.
Few remarks:
- why don't you use a `FlatPoint` to define Assemble size:

```
public class Assemble {

  /**
   * Root box width and height
   */
   private FlatPoint size;

  /**
   * All cells in root container
   */
  private List<Cell> cells;
}
```

- why a vertical_ratio and horizontal_ratio for the `Cell`? Couldn't the size of a cell be absolute?
- for us (PlantUML), label, Color, image link are completely useless. I suspect that it will be the same for every user of your library. That's what I mean when I say: **just to the layout, forget about the drawing**. It is your users that will code the drawing.
However, I understand that you need those information to manage the legacy HtmlLabel.
So I suggest that you group into a separate class all drawing information:

```
public class DrawingCell {

  /**
   * Label of cell
   */
  private String label;

  /**
   * This is an offset relative to the upper left corner of root box
   */
  private FlatPoint labelOffset;

  /**
   * Label font name
   */
  private String fontName;

  /**
   * Label font size
   */
  private FlatPoint fontSize;

  /**
   * Border width
   */
  private double penWidth = 1;

  /**
   * Image link
   */
  private String imgLink;

  /**
   * Border color
   */
  private Color color;

  /**
   * Font color
   */
  private Color fontColor;

  /**
   * Fill color
   */
  private Color fillColor;

  /**
   * Cell style
   */
  private NodeStyle style;

}
```
So your `Cell` class _could_ be simply:

```
public class Cell {

  /**
   * Cell Id
   */
  private String id;

  /**
   * The size of this cell is relative to the ratio of the root box
   */
  private FlatPoint size;

  /**
   * This an offset relative to the upper left corner of root box
   */
  private FlatPoint offset;


  /**
   * Cell shape (default is rectangle)
   */
  private NodeShape shape = NodeShapeEnum.RECT;

  /**
   * Drawing Information
   */
  private DrawingCell drawing;
 
}
```

And the more important: I guess that the library will allow to define edges from `Cell` and to `Cell`. right?

**Reply by jamisonjiang** [reply]
> why don't you use a FlatPoint to define Assemble size

If few attributes, I dont want to use a extra object.

> why a vertical_ratio and horizontal_ratio for the Cell? Couldn't the size of a cell be absolute?

Maybe it more convenient for html structure.

> However, I understand that you need those information to manage the legacy HtmlLabel.

Yeah, graph-support never drop the drawing, and this attributes it is important for user who only want to use the primative graphviz feature.

> And the more important: I guess that the library will allow to define edges from Cell and to Cell. right?

Right, use the tailCell and headCell to position endpoint of edge.

**Comment by jamisonjiang** [comment]
@arnaudroques I tried **Smetana**. I have never used it before. I thought there would be a big defect when I didn’t use it. But in fact, from a personal point of view, I don’t think it’s much different from the process of using dot, except for the cutting of some line segments. 

If **Smetana** is so successful that **PlantUML** is not satisfied, I think **graph-support** may not be more suitable for **PlantUML** than **Smetana**. After all, re-implementation means: just use a similar algorithm, the code has nothing to do with graphviz (this is different from VizJS and Smetana, these two projects are grapviz in a sense, but translated into other languages), so This will lead to a large difference between the final graph and graphviz, which will cause a large change in the style of the graph generated by **PlantUML**.
For example: the following is the rendering of the same graph data by graphviz and graph-support, and there is a big difference in the interval of processing the label of the cluster and the ordering of nodes. From the results of the algorithm, they are layered consistently, the cluster processing is consistent, and the number of intersections is consistent, but the results are not exactly the same.
![image](https://user-images.githubusercontent.com/125274994/230778510-fbfb5b7c-87c4-4672-959f-e34b0b7ec610.png)


So my conclusion is: **Smetana** is just an inaccurate line segment cutting, and **PlantUML** is not very acceptable. The completely subversive change of graph-support may be even more unacceptable to **PlantUML**. What do you think?

**Reply by arnaudroques** [reply]
> so This will lead to a large difference between the final graph and graphviz, which will cause a large change in the style of the graph generated by **PlantUML**. [...]
> they are layered consistently, the cluster processing is consistent, and the number of intersections is consistent, but the results are not exactly the same.

For us, it's not an issue: as you said, the result are _different_ but both are consistent. I would not say that one is better than the other.

We will probably let the end user choose which implementation he wants, as we did for [ELK by the way](https://plantuml.com/elk).

So that's fine.

**Reply by jamisonjiang** [reply]
> > so This will lead to a large difference between the final graph and graphviz, which will cause a large change in the style of the graph generated by **PlantUML**. [...]
> > they are layered consistently, the cluster processing is consistent, and the number of intersections is consistent, but the results are not exactly the same.
> 
> For us, it's not an issue: as you said, the result are _different_ but both are consistent. I would not say that one is better than the other.
> 
> We will probably let the end user choose which implementation he wants, as we did for [ELK by the way](https://plantuml.com/elk).
> 
> So that's fine.

Made some experimental attempts to determine the possibility of integration, not sure if it will work as expected, to see if further work is required.
https://github.com/jamisonjiang/plantuml

There is no precise calculation of the shape of the node, no setting of the coordinates of the label and the arrow, it is just a rough experiment.

**Comment by jamisonjiang** [comment]
@arnaudroques 1.1.0 supports the table feature and follows the previous idea.
**Assemble** uses the original relative coordinates to draw the Cell, and the **Table** provides the table layout of the Cell, and compiles the table into an **Assemble** for rendering. However, **Assemble** reuses the drawing of **Node** and the design of **label cells**. In fact, I did not spend too much to implement this feature, and the effect looks quite good! Thank you for your idea.

![image](https://github.com/plantuml/plantuml/assets/125274994/d2272caa-ae46-4f8a-91ac-f99f55b4155e)
![image](https://github.com/plantuml/plantuml/assets/125274994/21cf6eba-a0f4-4edd-ad86-92c85323cf49)
